lib/workflows/email_shutoff.py

```diff
@@ -0,0 +1,117 @@
+from typing import List, Optional, Dict, Tuple
+from openstack_query.api.query_objects import ServerQuery
+from enums.query.props.server_properties import ServerProperties
+from enums.query.props.user_properties import UserProperties
+from structs.email.email_params import EmailParams
+from enums.query.query_presets import QueryPresetsGeneric
+from email_api.emailer import Emailer
+from structs.email.smtp_account import SMTPAccount
+
+
+def query_shutoff_vms(project_id: Optional[List[str]] = None):
+    """
+    Sends an email to a user about a shutoff (stopped) VM
+    :param project_id: optional field to provide a list of project IDs for specific projects,
+    otherwise query all projects
+    :returns: Dictionary with a list of VM names grouped by user ID
+    """
+
+    query_obj = ServerQuery()
+
+    query_obj.select(
+        ServerProperties.SERVER_NAME
+    ).where(
+        QueryPresetsGeneric.EQUAL_TO,
+        ServerProperties.SERVER_STATUS,
+        value="SHUTOFF")
+
+    # if project ID provided as parameter
+    if project_id:
+        query_obj.run(
+            "openstack",
+            as_admin=True,
+            from_projects=project_id
+        )
+
+    else:
+        query_obj.run(
+            "openstack",
+            as_admin=True,
+            all_projects=True
+        )
+
+    user_query = query_obj.then("USER_QUERY", keep_previous_results=True)
+
+    user_query.select(
+        UserProperties.USER_EMAIL
+    ).group_by(
+        UserProperties.USER_NAME
+    )
+
+    if project_id:
+        user_query.run(
+            "openstack",
+            # from_projects=project_id
+        )
+
+    else:
+        user_query.run("openstack", as_admin=True, all_projects=True)
```

Comment:
you don't need to provide as_admin=True or all_projects=True for user_query. They are ignored here anyway

Reply:
I think this can be changed simply to 
```
    user_query.select(
        UserProperties.USER_EMAIL
    ).group_by(
        UserProperties.USER_NAME
    )
    user_query.run("openstack")
```

(As per the previous comments, the hardcoded "openstack" needs removing too